goal_src/engine/data/res.gc

```diff
@@ -402,13 +429,184 @@
+   default is the default result returned in the case of an error.
+   tag-addr is an address to a res-tag. The current base tag is written to this. Ignored if tag-addr is #f
+   buf-addr is an address to the data buffer used to write interpolated data to. It must have enough space! Only necessary for 'interp mode."
+
+  (let ((tag-pair (lookup-tag-idx obj name mode time)))
+      (cond
+        ((< (the-as int tag-pair) 0)
+         (empty)
+         )
+        (else
+          (let* ((tag (-> obj tag (-> tag-pair lo)))
+                 (tag-type (-> tag elt-type))
+                 (data (make-property-data obj time tag-pair buf-addr))
+                 )
+              (if tag-addr
+                (set! (-> tag-addr 0) tag)
+                )
+              (cond
+                ((type-type? tag-type uinteger)
+                 (case (-> tag elt-type size)
+                   ((1) (set! default (the-as uint128 (deref uint8 data))))
+                   ((2) (set! default (the-as uint128 (deref uint16 data))))
+                   ((4) (set! default (the-as uint128 (deref uint32 data))))
+                   ((16) (set! default (the-as uint128 (deref uint128 data))))
+                   (else (set! default (the-as uint128 (deref uint64 data))))
+                   )
+                 )
+                ((type-type? tag-type integer)
+                 (case (-> tag elt-type size)
+                   ((1) (set! default (the-as uint128 (deref int8 data))))
+                   ((2) (set! default (the-as uint128 (deref int16 data))))
+                   ((4) (set! default (the-as uint128 (deref int32 data))))
+                   ((16) (set! default (the-as uint128 (deref uint128 data))))
+                   (else (set! default (the-as uint128 (deref uint64 data))))
+                   )
+                 )
+                ((type-type? tag-type float)
+                 (set! default (the-as uint128 (deref float data)))
+                 )
+                (else
+                  (empty)
+                  )
+                )
+              )
+          )
+        )
+      )
+  default
   )
 
+(defmethod get-property-value2 res-lump ((obj res-lump) (name symbol) (mode symbol) (time float) (default uint128) (tag-addr (pointer res-tag)) (buf-addr pointer))
+  "same as get-property-value but float type is checked first?"
+
+  (let ((tag-pair (lookup-tag-idx obj name mode time)))
+      (cond
+        ((< (the-as int tag-pair) 0)
+         (empty)
+         )
+        (else
+          (let* ((tag (-> obj tag (-> tag-pair lo)))
+                 (tag-type (-> tag elt-type))
+                 (data (make-property-data obj time tag-pair buf-addr))
+                 )
+              (if tag-addr
+                (set! (-> tag-addr 0) tag)
+                )
+              (cond
+                ((type-type? tag-type float)
+                 (set! default (the-as uint128 (deref float data)))
```

Comment:
I want to double check the code generated for this case and make sure it matches GOAL in terms of what happens in the upper bits. I believe the correct behavior is that bits 32-64 are sign extended and 64-128 are 0.  But I suspect that bits 32-64 will end up zero in OpenGOAL.  I'm not sure if it really matters - the only thing I think of is comparing `float`s as integers will behave differently.

Reply:
Reading the disassembly is a bit too complex for me, but here is the disassembled output of the entire `res` file:
[test_disasm.txt](https://github.com/water111/jak-project/files/6580027/test_disasm.txt)

Reply:
According to the Intel manuals (and if I'm reading right), `movss` zero-extends bits 32 to 127 when reading from memory, but leaves them unmodified if reading from registers, so the end result for the `set!` is the latter, but its return value is zero-extended. Other than that, everything seems fine.

Reply:
Thanks! I took a look and the 128-bit stuff is working as expected. It's correctly treating `default` as a 128-bit variable and doing all the conversions as expected.  As written, if the size is 16, it will return a 128-bit value.  I'm not 100% sure that the original GOAL code does this correctly - there may have been a GOAL compiler bug that causes the upper 64-bits to be lost due to `or` being used instead of `por` in some cases.  The decompiler should eventually catch this for us (https://github.com/water111/jak-project/issues/549) so we can know for sure.

There are a few very minor issues (https://github.com/water111/jak-project/issues/548 https://github.com/water111/jak-project/issues/550), but I think we should merge this PR first and deal with those later.